src/main/scala/io/moia/protos/teleproto/FormatImpl.scala

```diff
@@ -71,42 +72,42 @@ trait FormatImpl {
   }
 
   /** From type `S[T]` extracts `T`. */
-  private[teleproto] def innerType(from: Type): Type =
+  private[teleproto] def innerType[T](from: Type[T]): Type[T] =
     from.typeArgs.headOption.getOrElse(abort(s"Type $from does not have type arguments"))
 
   /** Fails if types are not a Protobuf case class and case class pair. */
-  private[teleproto] def ensureValidTypes(protobufType: Type, modelType: Type): Unit =
+  private[teleproto] def ensureValidTypes[P, M](protobufType: Type[P], modelType: Type[M])(using Quotes): Unit =
     if (!checkClassTypes(protobufType, modelType)) {
       import scala.compiletime.error
       error(s"`$protobufType` and `$modelType` have to be case classes with a single parameter list!")
     }
 
-  private[teleproto] def checkClassTypes(protobufType: Type, modelType: Type): Boolean =
+  private[teleproto] def checkClassTypes[P, M](protobufType: Type[P], modelType: Type[M])(using Quotes): Boolean =
     isProtobuf(protobufType) && isSimpleCaseClass(modelType)
 
-  private[teleproto] def checkTraitTypes(protobufType: Type, modelType: Type): Boolean =
+  private[teleproto] def checkTraitTypes[P, M](protobufType: Type[P], modelType: Type[M])(using Quotes): Boolean =
     isSealedTrait(protobufType) && isSealedTrait(modelType)
 
   /** A sealed trait with that is a subtype of [[GeneratedOneof]]. */
-  private[teleproto] def checkHierarchyTypes(protobufType: Type, modelType: Type): Boolean =
+  private[teleproto] def checkHierarchyTypes[P, M](protobufType: Type[P], modelType: Type[M])(using Quotes): Boolean =
     isSealedTrait(modelType) && isSealedTrait(protobufType) && protobufType <:< typeOf[GeneratedOneof]
 
   /** A ScalaPB enumeration can be mapped to a detached sealed trait with corresponding case objects and vice versa. */
-  private[teleproto] def checkEnumerationTypes(protobufType: Type, modelType: Type): Boolean =
+  private[teleproto] def checkEnumerationTypes[P, M](protobufType: Type[P], modelType: Type[M])(using Quotes): Boolean =
     isScalaPBEnumeration(protobufType) && isSealedTrait(modelType)
 
-  private[teleproto] def isProtobuf(tpe: Type): Boolean =
+  private[teleproto] def isProtobuf[P](tpe: Type[P])(using Quotes): Boolean =
     isSimpleCaseClass(tpe) // && tpe <:< typeOf[GeneratedMessage]
 
-  private[teleproto] def isSimpleCaseClass(tpe: Type): Boolean =
-    tpe.typeSymbol match {
-      case cs: ClassSymbol =>
-        cs.isCaseClass && {
-          val cons = tpe.member(termNames.CONSTRUCTOR).asMethod
-          cons.paramLists.lengthCompare(1) == 0 // has a single parameter list
-        }
-      case _ => false
+  private[teleproto] def isSimpleCaseClass[T](tpe: Type[T])(using Quotes): Boolean = {
+    import quotes.reflect.*
+
+    val tpeSym = TypeTree.of[T].symbol
```

Comment:
This is how to get the `symbol` in Scala 3. The `tpe` could also be `given` I think, as we don't access the parameter directly.

Reply:
I think you're right. I think we can just specify `[T: Type]` in the type parameterisation? I think we're acutally going to be using this `tpeSim` most of the time instead, right?